search/mongo_search.go

```diff
@@ -96,10 +97,63 @@ func (m *MongoSearcher) createQuery(query Query, withOptions bool) *mgo.Query {
 // are used (since CreateQuery can't support joins).
 func (m *MongoSearcher) CreatePipeline(query Query) *mgo.Pipe {
 	c := m.db.C(models.PluralizeLowerResourceName(query.Resource))
-	p := []bson.M{{"$match": m.createQueryObject(query)}}
+	params := query.Params()
 
+	// Separate out all the chained SearchParams thare are NOT embedded (point
+	// to a Reference in a different document instead of a Resource embedded in the
+	// same document). These can be searched using a $lookup in the pipeline.
+	searchParams := []SearchParam{}
+	chainedSearchParams := []ReferenceParam{}
+
+	for _, p := range params {
+		switch p := p.(type) {
+		case *ReferenceParam:
+			// test if p is embedded
+			switch r := p.Reference.(type) {
+			case ChainedQueryReference:
+				if !r.IsEmbedded {
+					chainedSearchParams = append(chainedSearchParams, *p)
+					continue
+				}
+			}
+		}
+		searchParams = append(searchParams, p)
+	}
+
+	p := []bson.M{{"$match": m.createQueryObjectFromParams(searchParams)}}
 	o := query.Options()
 
+	// Support for chained search. Chained search is a 2-step process:
+	// 1. $lookup all foreign documents referenced by the chained search parameter(s)
+	// 2. $match the fields/values in those referenced documents specified by each chained search parameter
+
+	for _, chainedParam := range chainedSearchParams {
+
+		// The key, value pair for a chained search parameter is something like:
+		// ("subject:Patient.gender", "male")
+		param, val := chainedParam.getQueryParamAndValue()
+		info := chainedParam.getInfo()
+		fields := strings.FieldsFunc(param, func(c rune) bool { return !unicode.IsLetter(c) }) // splits on any non-letter character (e.g. ":", ".")
+		from := models.PluralizeLowerResourceName(fields[1])
+		localField := strings.ToLower(info.Paths[0].Path) + ".referenceid"
+		as := "_" + strings.ToLower(info.Paths[0].Path)
+
+		p = append(p, bson.M{"$lookup": bson.M{
+			"from":         from,
+			"localField":   localField,
+			"foreignField": "_id",
+			"as":           as,
+		}})
+
+		// Immediately follow each $lookup with a $match to narrow down the result set as soon as possible
+		key := as + "." + fields[2]
+		p = append(p, bson.M{"$match": bson.M{
+			key: val,
+		}})
```

Comment:
I think this bit is wrong.  In this case, it works because the search param name (_gender_) happens to match the property name (_gender_) exactly.  But some search param names won't match their fields.  For example `subject:Patient.address-city` needs to map to `address.city` rather than `address-city`.  Another (more pronounced) example: `subject:Patient.death-date` needs to map to `deceasedDateTime` rather than `death-date`.